File: backend/app/services/appointment_service.py
from typing import List, Dict, Optional
from prisma import Prisma
from prisma.errors import RecordNotFoundError
from app.schemas.student_schemas import AppointmentCreateSchema, AppointmentSchema


class AppointmentService:
    @staticmethod
    async def create_appointment(
        prisma: Prisma, student_id: int, appointment_data: AppointmentCreateSchema
    ) -> AppointmentSchema:
        """
        Schedule a new appointment.

        Args:
            prisma (Prisma): Prisma client.
            student_id (int): ID of the student.
            appointment_data (AppointmentCreateSchema): Appointment details.

        Returns:
            AppointmentSchema: Created appointment details.

        Raises:
            ValueError: If appointment details are invalid or if an error occurs during creation.
        """
        # Validate appointment data
        if not appointment_data.provider_id or not appointment_data.date:
            raise ValueError("Provider ID and date are required.")

        try:
            # Create appointment in the database
            appointment = await prisma.appointment.create(
                data={
                    "student_id": student_id,
                    "provider_id": appointment_data.provider_id,
                    "date": appointment_data.date,
                    "status": "PENDING",
                }
            )
            return AppointmentSchema.from_orm(appointment)

        except Exception as e:
            raise ValueError(f"Error scheduling appointment: {str(e)}")

    @staticmethod
    async def cancel_appointment(
        prisma: Prisma, student_id: int, appointment_id: int
    ) -> Dict[str, str]:
        """
        Cancel an existing appointment.

        Args:
            prisma (Prisma): Prisma client.
            student_id (int): ID of the student.
            appointment_id (int): ID of the appointment to cancel.

        Returns:
            Dict[str, str]: Confirmation message.

        Raises:
            ValueError: If the appointment is not found or if an error occurs during cancellation.
        """
        try:
            # Find the appointment
            appointment = await prisma.appointment.find_unique(
                where={"id": appointment_id, "student_id": student_id}
            )

            if not appointment:
                raise ValueError("Appointment not found.")

            # Update appointment status to 'CANCELLED'
            await prisma.appointment.update(
                where={"id": appointment_id}, data={"status": "CANCELLED"}
            )

            return {"message": "Appointment cancelled successfully."}

        except RecordNotFoundError:
            raise ValueError("Appointment not found.")
        except Exception as e:
            raise ValueError(f"Error canceling appointment: {str(e)}")

    @staticmethod
    async def get_student_appointments(
        prisma: Prisma,
        student_id: int,
        status: Optional[str] = None,
        start_date: Optional[str] = None,
        end_date: Optional[str] = None,
    ) -> List[AppointmentSchema]:
        """
        Retrieve student's appointments.

        Args:
            prisma (Prisma): Prisma client.
            student_id (int): ID of the student.
            status (Optional[str]): Filter by appointment status.
            start_date (Optional[str]): Filter by start date.
            end_date (Optional[str]): Filter by end date.

        Returns:
            List[AppointmentSchema]: List of appointments for the student.

        Raises:
            ValueError: If an error occurs during retrieval.
        """
        try:
            # Build query with optional filters
            query = prisma.appointment.find_many(
                where={"student_id": student_id},
                order_by={"date": "asc"},  # Order by date
            )

            # Apply filters if provided
            if status:
                query = query.where({"status": status})

            if start_date:
                query = query.where({"date": {"gte": start_date}})

            if end_date:
                query = query.where({"date": {"lte": end_date}})

            appointments = await query

            return [AppointmentSchema.from_orm(appt) for appt in appointments]

        except Exception as e:
            raise ValueError(f"Error retrieving appointments: {str(e)}")

File: backend/src/services/AppointmentService.js
const { Appointment } = require("../models");
const { ApiError } = require("../utils/apiError");
const logger = require("../utils/logger");

class AppointmentService {
  // Create a new appointment
  async createAppointment(studentId, appointmentData) {
    try {
      logger.info(`Creating appointment for student ${studentId}`);

      return await Appointment.createAppointment({
        studentId,
        service: appointmentData.service,
        startTime: new Date(appointmentData.startTime),
        duration: Number(appointmentData.duration),
        notes: appointmentData.notes,
        status: "PENDING",
      });
    } catch (error) {
      logger.error("Failed to create appointment:", error);
      throw new ApiError(500, "Failed to create appointment");
    }
  }

  // Fetch all appointments for a student
  async getAppointmentsByStudentId(studentId) {
    try {
      logger.info(`Fetching appointments for studentId: ${studentId}`);

      const appointments = await Appointment.getAppointmentsByStudentId(
        studentId
      );

      logger.info(
        `Fetched ${appointments.length} appointments for studentId: ${studentId}`
      );
      return appointments;
    } catch (error) {
      logger.error("Failed to fetch appointments:", error);
      throw new ApiError(500, "Failed to fetch appointments");
    }
  }

  // Fetch a single appointment by ID
  // async getAppointmentById(appointmentId) {
  //   try {
  //     const appointment = await Appointment.getAppointmentsByStudentId(
  //       appointmentId
  //     );
  //     if (!appointment) {
  //       throw new ApiError(404, "Appointment not found");
  //     }
  //     return appointment;
  //   } catch (error) {
  //     logger.error("Failed to fetch appointment:", error);
  //     throw new ApiError(500, "Failed to fetch appointment");
  //   }
  // }
  async getAppointmentById(appointmentId) {
    try {
      logger.info(
        `[Appointment Service] Fetching appointment ${appointmentId}`
      );

      const appointment = await Appointment.getAppointmentById(appointmentId);

      if (!appointment) {
        logger.warn(
          `[Appointment Service] Appointment ${appointmentId} not found`
        );
        throw new ApiError(404, "Appointment not found");
      }

      logger.info(
        `[Appointment Service] Successfully retrieved appointment ${appointmentId}`
      );
      return appointment;
    } catch (error) {
      logger.error(
        `[Appointment Service] Error getting appointment ${appointmentId}:`,
        error
      );

      // Preserve existing error status if available
      if (error instanceof ApiError) throw error;
      throw new ApiError(500, "Failed to retrieve appointment");
    }
  }
  // Update an appointment
  async updateAppointment(appointmentId, updateData) {
    try {
      const appointment = await Appointment.updateAppointment(
        appointmentId,
        updateData
      );
      if (!appointment) {
        throw new ApiError(404, "Appointment not found");
      }
      return appointment;
    } catch (error) {
      logger.error("Failed to update appointment:", error);
      throw new ApiError(500, "Failed to update appointment");
    }
  }

  // Delete an appointment
  async deleteAppointment(appointmentId) {
    try {
      const appointment = await Appointment.deleteAppointment(appointmentId);
      if (!appointment) {
        throw new ApiError(404, "Appointment not found");
      }
      return { message: "Appointment deleted successfully" };
    } catch (error) {
      logger.error("Failed to delete appointment:", error);
      throw new ApiError(500, "Failed to delete appointment");
    }
  }

  // Cancel an appointment
  async cancelAppointment(appointmentId) {
    try {
      const appointment = await Appointment.updateStatus(
        appointmentId,
        "CANCELLED"
      );
      if (!appointment) {
        throw new ApiError(404, "Appointment not found");
      }
      return { message: "Appointment cancelled successfully" };
    } catch (error) {
      logger.error("Failed to cancel appointment:", error);
      throw new ApiError(500, "Failed to cancel appointment");
    }
  }
  async rescheduleAppointment(appointmentId, updateData) {
    try {
      logger.info(`Rescheduling appointment ${appointmentId}`);
      console.log("[Service] Rescheduling appointment:", appointmentId);
      console.log("[Service] Update data:", updateData);
      const updatedAppointment = await Appointment.reschedule(
        appointmentId,
        updateData
      );

      if (!updatedAppointment) {
        console.error("[Service] Appointment not found:", appointmentId);
        throw new ApiError(404, "Appointment not found");
      }

      logger.info(`Appointment ${appointmentId} rescheduled successfully`);
      console.log(
        "[Service] Appointment rescheduled successfully:",
        updatedAppointment
      );
      return updatedAppointment;
    } catch (error) {
      logger.error("Failed to reschedule appointment:", error);
      throw new ApiError(500, "Failed to reschedule appointment");
    }
  }
  async countToday(providerId) {
    return Appointment.countToday(providerId);
  }

  async weeklyOverview(providerId) {
    return Appointment.weeklyOverview(providerId);
  }
}

module.exports = new AppointmentService();
